app/features/verified-programs/api.ts

```diff
@@ -0,0 +1,109 @@
+    const results = await Promise.allSettled(programIds.map(id => fetchProgramMetadata(id)));
+
+    const metadataMap = new Map<string, ProgramMetadata | null>();
+    results.forEach((result, index) => {
+        const programId = programIds[index];
+        metadataMap.set(programId, result.status === 'fulfilled' ? result.value : null);
+    });
+
+    return metadataMap;
+}
+
+// Fetch programs page by page with metadata
+export async function fetchProgramsProgressively(
+    onPageLoaded: (programs: VerifiedProgramInfo[], page: number, totalPages: number, totalCount: number) => void
+): Promise<void> {
+    try {
+        // Fetch first page to get total pages
+        const firstPageResponse = await fetch(`${API_BASE_URL}/list/1`);
+        if (!firstPageResponse.ok) {
+            throw new Error('Failed to fetch verified programs');
+        }
+        const firstPage: VerifiedProgramsResponse = await firstPageResponse.json();
+
+        const totalPages = firstPage.meta.total_pages;
+        const totalCount = firstPage.meta.total;
+
+        // Process first page
+        const firstPageMetadata = await fetchMetadataForPage(firstPage.verified_programs);
+        const firstPagePrograms = firstPage.verified_programs.map(programId => {
+            const metadata = firstPageMetadata.get(programId) ?? null;
+
+            return {
+                isVerified: true as const,
+                lastVerifiedAt: metadata?.last_verified_at,
+                name: getProgramName(programId, metadata?.repo_url),
+                programId,
+                repoUrl: metadata?.repo_url,
+            };
+        });
+
+        onPageLoaded(firstPagePrograms, 1, totalPages, totalCount);
+
+        // Fetch remaining pages sequentially
+        for (let page = 2; page <= totalPages; page++) {
+            const pageResponse = await fetch(`${API_BASE_URL}/list/${page}`);
+            if (!pageResponse.ok) {
+                Logger.error(`Failed to fetch page ${page}`);
+                continue;
+            }
+
+            const pageData: VerifiedProgramsResponse = await pageResponse.json();
+            const pageMetadata = await fetchMetadataForPage(pageData.verified_programs);
+            const pagePrograms = pageData.verified_programs.map(programId => {
+                const metadata = pageMetadata.get(programId) ?? null;
+
+                return {
+                    isVerified: true as const,
+                    lastVerifiedAt: metadata?.last_verified_at,
+                    name: getProgramName(programId, metadata?.repo_url),
+                    programId,
+                    repoUrl: metadata?.repo_url,
+                };
+            });
+
+            onPageLoaded(pagePrograms, page, totalPages, totalCount);
+
+            // Small delay between pages to avoid rate limiting
+            if (page < totalPages) {
+                await new Promise(resolve => setTimeout(resolve, 100));
```

Comment:
What is a reason to preload all the pages progressively? That will create a spike of requests

Reply:
I'd recommend not to fetch pages all at once as well. This creates problems at present and will not work well in the future, given that the number of these pages is expected to increase.